Tutorials/code/vpg.py

```diff
@@ -0,0 +1,191 @@
+import numpy as np
+import cv2
+import matplotlib.pyplot as plt
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.distributions import Categorical
+
+SIZE = 10
+HM_EPISODES = 25000
+MOVE_PENALTY = 1
+ENEMY_PENALTY = 300
+FOOD_REWARD = 25
+SHOW_EVERY = 1000
+
+PLAYER_N = 1
+FOOD_N = 2
+ENEMY_N = 3
+
+d = {1: (255, 175, 0),
+     2: (0, 255, 0),
+     3: (0, 0, 255)}
+
+CELL_SIZE = 30
+
+class Blob:
+    def __init__(self):
+        self.x = np.random.randint(0, SIZE)
+        self.y = np.random.randint(0, SIZE)
+
+    def __str__(self):
+        return f"{self.x}, {self.y}"
+
+    def __sub__(self, other):
+        return (self.x-other.x, self.y-other.y)
+
+    def action(self, choice):
+        if choice == 0:
+            self.move(x=1, y=1)
+        elif choice == 1:
+            self.move(x=-1, y=-1)
+        elif choice == 2:
+            self.move(x=-1, y=1)
+        elif choice == 3:
+            self.move(x=1, y=-1)
+
+    def move(self, x=False, y=False):
+        if not x:
+            self.x += np.random.randint(-1, 2)
+        else:
+            self.x += x
+        if not y:
+            self.y += np.random.randint(-1, 2)
+        else:
+            self.y += y
+
+        if self.x < 0:
+            self.x = 0
+        elif self.x > SIZE-1:
+            self.x = SIZE-1
+        if self.y < 0:
+            self.y = 0
+        elif self.y > SIZE-1:
+            self.y = SIZE-1
+
+def get_state(player, food, enemy):
+    return (player-food) + (player-enemy)
+
+def render_env(player, food, enemy):
+    env = np.zeros((SIZE*CELL_SIZE, SIZE*CELL_SIZE, 3), dtype=np.uint8)
+    
+    # Draw grid lines
+    env[:, ::CELL_SIZE] = [100, 100, 100]  # Vertical lines
+    env[::CELL_SIZE, :] = [100, 100, 100]  # Horizontal lines
+    
+    # Draw food
+    env[food.y*CELL_SIZE:(food.y+1)*CELL_SIZE, food.x*CELL_SIZE:(food.x+1)*CELL_SIZE] = d[FOOD_N]
+    
+    # Draw player
+    env[player.y*CELL_SIZE:(player.y+1)*CELL_SIZE, player.x*CELL_SIZE:(player.x+1)*CELL_SIZE] = d[PLAYER_N]
+    
+    # Draw enemy
+    env[enemy.y*CELL_SIZE:(enemy.y+1)*CELL_SIZE, enemy.x*CELL_SIZE:(enemy.x+1)*CELL_SIZE] = d[ENEMY_N]
+    
+    return env
+
+class PolicyNetwork(nn.Module):
+    def __init__(self, input_size, output_size):
+        super(PolicyNetwork, self).__init__()
+        self.fc = nn.Sequential(
+            nn.Linear(input_size, 64),
+            nn.ReLU(),
+            nn.Linear(64, 64),
+            nn.ReLU(),
+            nn.Linear(64, output_size)
+        )
+
+    def forward(self, x):
+        return torch.softmax(self.fc(x), dim=-1)
+
+def select_action(state):
+    state = torch.FloatTensor(state).unsqueeze(0)
+    probs = policy_net(state)
+    m = Categorical(probs)
+    action = m.sample()
+    return action.item(), m.log_prob(action)
+
+def update_policy(log_probs, rewards):
```

Comment:
Looks good just need to add more comments for each defined variable

Also, a gif would be better for the agents